modules/log/logger.go

```diff
@@ -4,149 +4,140 @@
-	}
-
-	l.MultiChannelledLog.DelLogger(name)
-
-	err = l.MultiChannelledLog.AddLogger(eventLogger)
-	if err != nil {
-		if IsErrDuplicateName(err) {
-			return fmt.Errorf("Duplicate named sublogger %s %v", name, l.MultiChannelledLog.GetEventLoggerNames())
-		}
-		return fmt.Errorf("Failed to add sublogger (%s): %v", name, err)
-	}
-
-	return nil
-}
-
-// DelLogger deletes a sublogger from this logger.
-func (l *Logger) DelLogger(name string) (bool, error) {
-	return l.MultiChannelledLog.DelLogger(name), nil
-}
-
-// Log msg at the provided level with the provided caller defined by skip (0 being the function that calls this function)
-func (l *Logger) Log(skip int, level Level, format string, v ...interface{}) error {
-	if l.GetLevel() > level {
-		return nil
-	}
-	caller := "?()"
-	pc, filename, line, ok := runtime.Caller(skip + 1)
-	if ok {
-		// Get caller function name.
-		fn := runtime.FuncForPC(pc)
-		if fn != nil {
-			caller = fn.Name() + "()"
-		}
-	}
-	msg := format
-	if len(v) > 0 {
-		msg = ColorSprintf(format, v...)
-	}
-	stack := ""
-	if l.GetStacktraceLevel() <= level {
-		stack = Stack(skip + 1)
-	}
-	return l.SendLog(level, caller, strings.TrimPrefix(filename, prefix), line, msg, stack)
-}
-
-// SendLog sends a log event at the provided level with the information given
-func (l *Logger) SendLog(level Level, caller, filename string, line int, msg string, stack string) error {
-	if l.GetLevel() > level {
-		return nil
-	}
-	event := &Event{
-		level:      level,
-		caller:     caller,
-		filename:   filename,
-		line:       line,
-		msg:        msg,
-		time:       time.Now(),
-		stacktrace: stack,
-	}
-	l.LogEvent(event)
-	return nil
+import "os"
+
+// Logger is the basic interface for logging
+type Logger interface {
+	LevelLogger
+	Trace(format string, v ...interface{})
+	IsTrace() bool
+	Debug(format string, v ...interface{})
+	IsDebug() bool
+	Info(format string, v ...interface{})
+	IsInfo() bool
+	Warn(format string, v ...interface{})
+	IsWarn() bool
+	Error(format string, v ...interface{})
+	ErrorWithSkip(skip int, format string, v ...interface{})
+	IsError() bool
+	Critical(format string, v ...interface{})
+	CriticalWithSkip(skip int, format string, v ...interface{})
+	IsCritical() bool
+	Fatal(format string, v ...interface{})
+	FatalWithSkip(skip int, format string, v ...interface{})
+	IsFatal() bool
+}
+
+// LevelLogger is the simplest logging interface
+type LevelLogger interface {
+	Flush()
+	Close()
+	GetLevel() Level
+	Log(skip int, level Level, format string, v ...interface{}) error
+}
+
+// SettableLogger is the interface of loggers which have subloggers
+type SettableLogger interface {
+	SetLogger(name, provider, config string) error
+	DelLogger(name string) (bool, error)
+}
+
+// StacktraceLogger is a logger that can log stacktraces
+type StacktraceLogger interface {
+	GetStacktraceLevel() Level
+}
+
+// LevelLoggerLogger wraps a LevelLogger as a Logger
+type LevelLoggerLogger struct {
```

Comment:
I'm not sure if this is the best name for it, but I do not have a better idear

Reply:
It's a bit of a J2EE name I guess... But it does exactly what it says it does